# Eerste verkenning en validatie van de observation details. 
Verkennen van de data en uitvoeren van een eerste ruwe cleaning. </br>
Het resultaat wordt geplaatst in de sectie 'cleaned'

In [21]:
import pandas as pd
import os

In [22]:
species_id =116
folder_path_clean = "./scraped_data/cleaned/"

In [23]:
os.getcwd()
# set the max columns to none
pd.set_option('display.max_columns', None)
# set the max columns to none
pd.set_option('display.max_rows', None)


## Verkenning van de data


In [24]:
# file path
observation_details = f'scraped_data/observation_details_{species_id}_complete.csv'


# Load the data
df = pd.read_csv(observation_details, skipinitialspace=True, index_col=0)
df.head()

species_id     species_name species_name_scientific  \
observation_id                                                        
336301369            116.0  Halsbandparkiet      Psittacula krameri   
336233189            116.0  Halsbandparkiet      Psittacula krameri   
336205268            116.0  Halsbandparkiet      Psittacula krameri   
336194978            116.0  Halsbandparkiet      Psittacula krameri   
336193018            116.0  Halsbandparkiet      Psittacula krameri   

                                          validation  gps_coordinates  \
observation_id                                                          
336301369       Goedgekeurd (automatische validatie)  50.8585, 4.6717   
336233189       Goedgekeurd (automatische validatie)  50.8044, 4.6010   
336205268       Goedgekeurd (automatische validatie)  50.8584, 4.2921   
336194978       Goedgekeurd (automatische validatie)  50.9315, 4.3675   
336193018       Goedgekeurd (automatische validatie)  50.7396, 3.3351   

               accuracy       source              date amount life_stage  \
observation_id                                                             
336301369           33m         iObs  2024-12-31 16:45      1   onbekend   
336233189           30m     Showcase  2024-12-31 16:00      1   onbekend   
336205268            3m      ObsMapp  2024-12-31 15:29      6   onbekend   
336194978           13m  ObsIdentify  2024-12-31 15:21      1   onbekend   
336193018           14m         Site  2024-12-31 14:57      4   onbekend   

                   activity  location_id  \
observation_id                             
336301369       ter plaatse      30919.0   
336233189       ter plaatse      30976.0   
336205268       foeragerend      32596.0   
336194978       ter plaatse      23260.0   
336193018       ter plaatse      27358.0   

                                                         location  \
observation_id                                                      
336301369               Heverlee - Egenhoven - Egenhoven Bos (VB)   
336233189                              Loonbeek - Korenheide (VB)   
336205268       Berchem-Ste-Agathe/St-Agatha-Berchem - Koninck...   
336194978                                         Grimbergen (VB)   
336193018                                            Kooigem (WV)   

                observer_id         observer_name counting_method  \
observation_id                                                      
336301369           40236.0        Frederik Fluyt        onbekend   
336233189            3528.0          Externe bron            geen   
336205268           72841.0           Sara Velghe        onbekend   
336194978          359678.0      Kristof Vynckier            geen   
336193018          201247.0  Lore Vander Plancken            geen   

                           method  
observation_id                     
336301369                onbekend  
336233189                onbekend  
336205268       gezien en gehoord  
336194978                onbekend  
336193018                  gezien

In [25]:
df.index 

Index([336301369, 336233189, 336205268, 336194978, 336193018, 336289624,
       336284264, 336284263, 336284262, 336204005,
       ...
       233647277, 233647261, 233644993,  98988761, 145469911,  50541272,
       203905722,  50531480, 142655244, 233392391],
      dtype='int64', name='observation_id', length=96951)

In [26]:
df.columns

Index(['species_id', 'species_name', 'species_name_scientific', 'validation',
       'gps_coordinates', 'accuracy', 'source', 'date', 'amount', 'life_stage',
       'activity', 'location_id', 'location', 'observer_id', 'observer_name',
       'counting_method', 'method'],
      dtype='object')

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 96951 entries, 336301369 to 233392391
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   species_id               96950 non-null  float64
 1   species_name             96950 non-null  object 
 2   species_name_scientific  96950 non-null  object 
 3   validation               96950 non-null  object 
 4   gps_coordinates          96551 non-null  object 
 5   accuracy                 96493 non-null  object 
 6   source                   92767 non-null  object 
 7   date                     96950 non-null  object 
 8   amount                   96950 non-null  object 
 9   life_stage               96575 non-null  object 
 10  activity                 96575 non-null  object 
 11  location_id              96092 non-null  float64
 12  location                 96092 non-null  object 
 13  observer_id              96793 non-null  float64
 14  observer_name  

In [28]:
stats = df.describe(include='all')
stats

species_id     species_name species_name_scientific  \
count      96950.0            96950                   96950   
unique         NaN                1                       1   
top            NaN  Halsbandparkiet      Psittacula krameri   
freq           NaN            96950                   96950   
mean         116.0              NaN                     NaN   
std            0.0              NaN                     NaN   
min          116.0              NaN                     NaN   
25%          116.0              NaN                     NaN   
50%          116.0              NaN                     NaN   
75%          116.0              NaN                     NaN   
max          116.0              NaN                     NaN   

                                  validation  gps_coordinates accuracy source  \
count                                  96950            96551    96493  92767   
unique                                     5            63817      378     26   
top     Goedgekeurd (automatische validatie)  51.0509, 4.3699      10m   Site   
freq                                   57686              961    24830  46331   
mean                                     NaN              NaN      NaN    NaN   
std                                      NaN              NaN      NaN    NaN   
min                                      NaN              NaN      NaN    NaN   
25%                                      NaN              NaN      NaN    NaN   
50%                                      NaN              NaN      NaN    NaN   
75%                                      NaN              NaN      NaN    NaN   
max                                      NaN              NaN      NaN    NaN   

              date amount life_stage     activity    location_id  \
count        96950  96950      96575        96575   96092.000000   
unique       63867    556         19           44            NaN   
top     2022-11-15      1   onbekend  ter plaatse            NaN   
freq            34  46089      88135        56295            NaN   
mean           NaN    NaN        NaN          NaN   60541.305551   
std            NaN    NaN        NaN          NaN  101529.453801   
min            NaN    NaN        NaN          NaN   23088.000000   
25%            NaN    NaN        NaN          NaN   24033.000000   
50%            NaN    NaN        NaN          NaN   29069.000000   
75%            NaN    NaN        NaN          NaN   41806.000000   
max            NaN    NaN        NaN          NaN  724090.000000   

               location   observer_id        observer_name counting_method  \
count             96092  9.679300e+04                96793           96950   
unique             3090           NaN                 6037               7   
top     Willebroek (AN)           NaN  Wim Van den Bossche        onbekend   
freq               7041           NaN                14471           79414   
mean                NaN  8.840309e+04                  NaN             NaN   
std                 NaN  1.131419e+05                  NaN             NaN   
min                 NaN  1.850000e+02                  NaN             NaN   
25%                 NaN  4.042000e+04                  NaN             NaN   
50%                 NaN  4.645600e+04                  NaN             NaN   
75%                 NaN  9.408200e+04                  NaN             NaN   
max                 NaN  1.058369e+06                  NaN             NaN   

          method  
count      96950  
unique        15  
top     onbekend  
freq       60871  
mean         NaN  
std          NaN  
min          NaN  
25%          NaN  
50%          NaN  
75%          NaN  
max          NaN

In [29]:
df.isnull().sum()

species_id                    1
species_name                  1
species_name_scientific       1
validation                    1
gps_coordinates             400
accuracy                    458
source                     4184
date                          1
amount                        1
life_stage                  376
activity                    376
location_id                 859
location                    859
observer_id                 158
observer_name               158
counting_method               1
method                        1
dtype: int64

# First cleaning

## Ongeldige species_id (NaN)

In [30]:
df[df['species_id'].isnull()]

species_id species_name species_name_scientific validation  \
observation_id                                                               
322156271              NaN          NaN                     NaN        NaN   

               gps_coordinates accuracy source date amount life_stage  \
observation_id                                                          
322156271                  NaN      NaN    NaN  NaN    NaN        NaN   

               activity  location_id location  observer_id observer_name  \
observation_id                                                             
322156271           NaN          NaN      NaN          NaN           NaN   

               counting_method method  
observation_id                         
322156271                  NaN    NaN

De waarnemingen (observation_id) met species_id Nan  

Halsbandparkiet (species_id = 116)  
- 322156271  

Boomklever (species_id = 70)  
- 211495505  
- 210627812  
- 210627870  
- 203114064  

Als we deze verifiëren op de website eg. (https://waarnemingen.be/observation/322156271/), dan zien we dat deze inderdaad niet bestaan.  
Een verklaring hiervoor is dat deze nog wel bestonden toen we de general data ophaalden, maar nadien verwijderd zijn geweest. Vermits deze waarnemingen niet meer bestaan, verwijderen we deze ook uit onze dataset.

In [31]:
df_first_cleaning = df.dropna(subset=['species_id'])

## Species_id
De species_id is niet degene die we verwachten. Bij de general observations hebben we op een specifieke soort gezocht, maar nu blijkt dat bij de observation_details de boomklever, dat deze niet altijd overeenkomt met degene die we wilden ophalen. </br>

#### conclusie
Bij de boomklevers zien we dat we ook een species_id 258977 en 258978 hebben. Dit blijken ondersoorten van de boomklever te zijn , nl Boomklever ssp caesia en Boomklever ssp europaea </br>
In onze oefening gaan we het onderscheid tussen deze soorten niet maken, deze ondersoort blijft behouden in de data.

In [32]:
df_first_cleaning.groupby('species_id').nunique()

species_name  species_name_scientific  validation  \
species_id                                                      
116.0                  1                        1           5   

            gps_coordinates  accuracy  source   date  amount  life_stage  \
species_id                                                                 
116.0                 63817       378      26  63867     556          19   

            activity  location_id  location  observer_id  observer_name  \
species_id                                                                
116.0             44         3095      3090         6151           6037   

            counting_method  method  
species_id                           
116.0                     7      15

In [33]:
df1 = df_first_cleaning[df_first_cleaning['species_id'] == 70].head(2)
df2 = df_first_cleaning[df_first_cleaning['species_id'] == 258977.0].head(2)
df3 = df_first_cleaning[df_first_cleaning['species_id'] == 258978.0].head(2)

pd.concat([df1, df2, df3], axis=0)

Empty DataFrame
Columns: [species_id, species_name, species_name_scientific, validation, gps_coordinates, accuracy, source, date, amount, life_stage, activity, location_id, location, observer_id, observer_name, counting_method, method]
Index: []

In [34]:
stats = df_first_cleaning.describe(include='all')
stats

species_id     species_name species_name_scientific  \
count      96950.0            96950                   96950   
unique         NaN                1                       1   
top            NaN  Halsbandparkiet      Psittacula krameri   
freq           NaN            96950                   96950   
mean         116.0              NaN                     NaN   
std            0.0              NaN                     NaN   
min          116.0              NaN                     NaN   
25%          116.0              NaN                     NaN   
50%          116.0              NaN                     NaN   
75%          116.0              NaN                     NaN   
max          116.0              NaN                     NaN   

                                  validation  gps_coordinates accuracy source  \
count                                  96950            96551    96493  92767   
unique                                     5            63817      378     26   
top     Goedgekeurd (automatische validatie)  51.0509, 4.3699      10m   Site   
freq                                   57686              961    24830  46331   
mean                                     NaN              NaN      NaN    NaN   
std                                      NaN              NaN      NaN    NaN   
min                                      NaN              NaN      NaN    NaN   
25%                                      NaN              NaN      NaN    NaN   
50%                                      NaN              NaN      NaN    NaN   
75%                                      NaN              NaN      NaN    NaN   
max                                      NaN              NaN      NaN    NaN   

              date amount life_stage     activity    location_id  \
count        96950  96950      96575        96575   96092.000000   
unique       63867    556         19           44            NaN   
top     2022-11-15      1   onbekend  ter plaatse            NaN   
freq            34  46089      88135        56295            NaN   
mean           NaN    NaN        NaN          NaN   60541.305551   
std            NaN    NaN        NaN          NaN  101529.453801   
min            NaN    NaN        NaN          NaN   23088.000000   
25%            NaN    NaN        NaN          NaN   24033.000000   
50%            NaN    NaN        NaN          NaN   29069.000000   
75%            NaN    NaN        NaN          NaN   41806.000000   
max            NaN    NaN        NaN          NaN  724090.000000   

               location   observer_id        observer_name counting_method  \
count             96092  9.679300e+04                96793           96950   
unique             3090           NaN                 6037               7   
top     Willebroek (AN)           NaN  Wim Van den Bossche        onbekend   
freq               7041           NaN                14471           79414   
mean                NaN  8.840309e+04                  NaN             NaN   
std                 NaN  1.131419e+05                  NaN             NaN   
min                 NaN  1.850000e+02                  NaN             NaN   
25%                 NaN  4.042000e+04                  NaN             NaN   
50%                 NaN  4.645600e+04                  NaN             NaN   
75%                 NaN  9.408200e+04                  NaN             NaN   
max                 NaN  1.058369e+06                  NaN             NaN   

          method  
count      96950  
unique        15  
top     onbekend  
freq       60871  
mean         NaN  
std          NaN  
min          NaN  
25%          NaN  
50%          NaN  
75%          NaN  
max          NaN

# Gebruikte waarden en hun frequentie checken

#### conclusie
- amount zal nog moeten worden opgekuist naderhand, want bevat niet enkel het aantal vogels, maar ook tekst
- de observatiedatums starten al van in 1930, we zullen moeten beslissen of we deze willen opnemen in onze dataset



In [35]:
# Kijken welke mogelijke waarden er zijn, en wat de frequentie in de andere kolommen is
df_first_cleaning.groupby('validation').nunique()

species_id  species_name  \
validation                                                       
(nog) niet te beoordelen                       1             1   
Goedgekeurd (aannemelijk)                      1             1   
Goedgekeurd (automatische validatie)           1             1   
Goedgekeurd (met bewijs)                       1             1   
Onbekend                                       1             1   

                                      species_name_scientific  \
validation                                                      
(nog) niet te beoordelen                                    1   
Goedgekeurd (aannemelijk)                                   1   
Goedgekeurd (automatische validatie)                        1   
Goedgekeurd (met bewijs)                                    1   
Onbekend                                                    1   

                                      gps_coordinates  accuracy  source  \
validation                                                                
(nog) niet te beoordelen                           71        19       5   
Goedgekeurd (aannemelijk)                        9981       152      15   
Goedgekeurd (automatische validatie)            38344       271      23   
Goedgekeurd (met bewijs)                         2639       120      12   
Onbekend                                        16280       192      19   

                                       date  amount  life_stage  activity  \
validation                                                                  
(nog) niet te beoordelen                 88      11           2        13   
Goedgekeurd (aannemelijk)              9471     202           8        32   
Goedgekeurd (automatische validatie)  39464     348          16        42   
Goedgekeurd (met bewijs)               3040      88           8        36   
Onbekend                              18306     181          13        31   

                                      location_id  location  observer_id  \
validation                                                                 
(nog) niet te beoordelen                       64        64           74   
Goedgekeurd (aannemelijk)                     894       892          778   
Goedgekeurd (automatische validatie)         1996      1992         4438   
Goedgekeurd (met bewijs)                      812       809         1390   
Onbekend                                     2314      2310         2659   

                                      observer_name  counting_method  method  
validation                                                                    
(nog) niet te beoordelen                         74                5       6  
Goedgekeurd (aannemelijk)                       773                6       7  
Goedgekeurd (automatische validatie)           4384                7      13  
Goedgekeurd (met bewijs)                       1378                7      10  
Onbekend                                       2649                7      11

In [36]:
df_first_cleaning.groupby('species_name').nunique()

species_id  species_name_scientific  validation  \
species_name                                                       
Halsbandparkiet           1                        1           5   

                 gps_coordinates  accuracy  source   date  amount  life_stage  \
species_name                                                                    
Halsbandparkiet            63817       378      26  63867     556          19   

                 activity  location_id  location  observer_id  observer_name  \
species_name                                                                   
Halsbandparkiet        44         3095      3090         6151           6037   

                 counting_method  method  
species_name                              
Halsbandparkiet                7      15

In [37]:
df_first_cleaning.groupby('amount').nunique().head(10)

species_id  species_name  \
amount                                                                         
1                                                            1             1   
1\n            \n            \n\n\n\n\n\n    \n...           1             1   
1\n            \n            \n\n\n\n\n\n    \n...           1             1   
1\n            \n            \n\n\n\n\n\n    \n...           1             1   
1\n            \n            \n\n\n\n\n\n    \n...           1             1   
1\n            \n            \n\n\n\n\n\n    \n...           1             1   
1\n            \n            \n\n\n\n\n\n    \n...           1             1   
1\n            \n            \n\n\n\n\n\n    \n...           1             1   
1\n            \n            \n\n\n\n\n\n    \n...           1             1   
1\n            \n            \n\n\n\n\n\n    \n...           1             1   

                                                    species_name_scientific  \
amount                                                                        
1                                                                         1   
1\n            \n            \n\n\n\n\n\n    \n...                        1   
1\n            \n            \n\n\n\n\n\n    \n...                        1   
1\n            \n            \n\n\n\n\n\n    \n...                        1   
1\n            \n            \n\n\n\n\n\n    \n...                        1   
1\n            \n            \n\n\n\n\n\n    \n...                        1   
1\n            \n            \n\n\n\n\n\n    \n...                        1   
1\n            \n            \n\n\n\n\n\n    \n...                        1   
1\n            \n            \n\n\n\n\n\n    \n...                        1   
1\n            \n            \n\n\n\n\n\n    \n...                        1   

                                                    validation  \
amount                                                           
1                                                            5   
1\n            \n            \n\n\n\n\n\n    \n...           1   
1\n            \n            \n\n\n\n\n\n    \n...           1   
1\n            \n            \n\n\n\n\n\n    \n...           1   
1\n            \n            \n\n\n\n\n\n    \n...           1   
1\n            \n            \n\n\n\n\n\n    \n...           2   
1\n            \n            \n\n\n\n\n\n    \n...           1   
1\n            \n            \n\n\n\n\n\n    \n...           1   
1\n            \n            \n\n\n\n\n\n    \n...           4   
1\n            \n            \n\n\n\n\n\n    \n...           2   

                                                    gps_coordinates  accuracy  \
amount                                                                          
1                                                             34082       305   
1\n            \n            \n\n\n\n\n\n    \n...                4         2   
1\n            \n            \n\n\n\n\n\n    \n...                1         1   
1\n            \n            \n\n\n\n\n\n    \n...                1         1   
1\n            \n            \n\n\n\n\n\n    \n...                2         2   
1\n            \n            \n\n\n\n\n\n    \n...                2         1   
1\n            \n            \n\n\n\n\n\n    \n...                1         1   
1\n            \n            \n\n\n\n\n\n    \n...                1         1   
1\n            \n            \n\n\n\n\n\n    \n...                8         6   
1\n            \n            \n\n\n\n\n\n    \n...                5         3   

                                                    source   date  life_stage  \
amount                                                                          
1                                                       26  35112          14   
1\n            \n            \n\n\n\n\n\n    \n...       1      3           0   
1\n            \n            \n\n\n\n\n\n    \n...  

In [38]:
df_first_cleaning.groupby('date').nunique().head(10)

species_id  species_name  species_name_scientific  validation  \
date                                                                        
1971-10-03           1             1                        1           1   
1972-01-08           1             1                        1           1   
1975-03-05           1             1                        1           1   
1975-05-01           1             1                        1           1   
1975-06-23           1             1                        1           1   
1977-05-07           1             1                        1           1   
1978-12-03           1             1                        1           1   
1979-01-26           1             1                        1           1   
1979-03-06           1             1                        1           1   
1979-03-15           1             1                        1           1   

            gps_coordinates  accuracy  source  amount  life_stage  activity  \
date                                                                          
1971-10-03                1         1       1       1           1         1   
1972-01-08                1         1       1       1           1         1   
1975-03-05                1         1       1       1           1         1   
1975-05-01                1         1       1       1           1         1   
1975-06-23                1         1       1       1           1         1   
1977-05-07                1         1       1       1           1         1   
1978-12-03                1         1       1       1           1         1   
1979-01-26                1         1       1       1           1         1   
1979-03-06                1         1       1       1           1         1   
1979-03-15                1         1       1       1           1         1   

            location_id  location  observer_id  observer_name  \
date                                                            
1971-10-03            1         1            1              1   
1972-01-08            1         1            1              1   
1975-03-05            1         1            1              1   
1975-05-01            1         1            1              1   
1975-06-23            0         0            1              1   
1977-05-07            1         1            1              1   
1978-12-03            1         1            1              1   
1979-01-26            1         1            1              1   
1979-03-06            1         1            1              1   
1979-03-15            1         1            1              1   

            counting_method  method  
date                                 
1971-10-03                1       1  
1972-01-08                1       1  
1975-03-05                1       1  
1975-05-01                1       1  
1975-06-23                1       1  
1977-05-07                1       1  
1978-12-03                1       1  
1979-01-26                1       1  
1979-03-06                1       1  
1979-03-15                1       1

In [39]:
df_first_cleaning[df_first_cleaning['date'] == '1946-01-01'].head(2)

Empty DataFrame
Columns: [species_id, species_name, species_name_scientific, validation, gps_coordinates, accuracy, source, date, amount, life_stage, activity, location_id, location, observer_id, observer_name, counting_method, method]
Index: []

## Saven van de gecleande data

In [40]:
# Save the cleaned DataFrame to a new CSV file
df_first_cleaning.to_csv(os.path.join(folder_path_clean, f"observation_details_{species_id}_clean.csv"), index=True)